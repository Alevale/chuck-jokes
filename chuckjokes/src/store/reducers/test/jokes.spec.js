import reducer, {buildFavoritesListFromServer} from '../jokes'
import {joke1, joke2} from './examples'

describe('Joke Reducer', () => {

    it('returns an initial state when passed undefined', () => {
        const result = reducer(undefined, { type: 'ANYTHING' })
        expect(result).toMatchSnapshot()
    })

    it('removes a panel when made to be toggled', () => {
        const startState = {
            expanded: [
                'panel1', 'panel2'
            ]
        }
        const action = { type: 'TOGGLE_PANEL', payload: 'panel1' }
        const result = reducer(startState, action)
        expect(result).toMatchSnapshot()
    })

    it('adds a favorite joke', () => {
        const startState = {
            jokes: [],
            favoriteJokes: []
        }
        const action = { type: 'ADD_FAVORITE_JOKE', payload: joke1 }
        const result = reducer(startState, action)
        expect(result).toMatchSnapshot()
    })

    it('adds a joke to the favorite list and removes it from the jokes list', () => {
        const startState = {
            jokes: [joke1],
            favoriteJokes: []
        }
        const action = { type: 'ADD_FAVORITE_JOKE', payload: joke1 }
        const result = reducer(startState, action)
        expect(result).toMatchSnapshot()
    })

    it('merges the local favorites with the user cached favorites', () => {
        const startState = {
            favoriteJokes: [joke1]
        }
        const userCache = {
            result: {
                joke: {
                    favoriteJokes: [joke2]
                }
            }
        }
        const dispatch = jest.fn()
        const getState = jest.fn()
        getState.mockReturnValueOnce({
            joke: startState
        })
        const result = buildFavoritesListFromServer(userCache)(dispatch, getState)
        expect(dispatch).toMatchSnapshot()
        expect(result).toMatchSnapshot()
    })

    it('merges the local favorites with the user cached favorites but does not create duplicates', () => {
        const startState = {
            favoriteJokes: [joke1]
        }
        const userCache = {
            result: {
                joke: {
                    favoriteJokes: [joke1]
                }
            }
        }
        const dispatch = jest.fn()
        const getState = jest.fn()
        getState.mockReturnValueOnce({
            joke: startState
        })
        const result = buildFavoritesListFromServer(userCache)(dispatch, getState)
        expect(dispatch).toMatchSnapshot()
        expect(result).toMatchSnapshot()
    })

})
