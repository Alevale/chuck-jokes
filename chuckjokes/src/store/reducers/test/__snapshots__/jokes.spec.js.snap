// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Joke Reducer adds a favorite joke 1`] = `
Object {
  "favoriteJokes": Array [
    Object {
      "categories": Array [],
      "id": 416,
      "joke": "Chuck Norris plays racquetball with a waffle iron and a bowling ball.",
    },
  ],
  "jokes": Array [],
}
`;

exports[`Joke Reducer adds a joke to the favorite list and removes it from the jokes list 1`] = `
Object {
  "favoriteJokes": Array [
    Object {
      "categories": Array [],
      "id": 416,
      "joke": "Chuck Norris plays racquetball with a waffle iron and a bowling ball.",
    },
  ],
  "jokes": Array [],
}
`;

exports[`Joke Reducer merges the local favorites with the user cached favorites 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "payload": Array [
          Object {
            "categories": Array [],
            "id": 416,
            "joke": "Chuck Norris plays racquetball with a waffle iron and a bowling ball.",
          },
          Object {
            "categories": Array [],
            "id": 33,
            "joke": "This is the 2 joke",
          },
        ],
        "type": "REPLACE_FAVORITE_JOKE",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Joke Reducer merges the local favorites with the user cached favorites 2`] = `undefined`;

exports[`Joke Reducer merges the local favorites with the user cached favorites but does not create duplicates 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "payload": Array [
          Object {
            "categories": Array [],
            "id": 416,
            "joke": "Chuck Norris plays racquetball with a waffle iron and a bowling ball.",
          },
        ],
        "type": "REPLACE_FAVORITE_JOKE",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Joke Reducer merges the local favorites with the user cached favorites but does not create duplicates 2`] = `undefined`;

exports[`Joke Reducer removes a panel when made to be toggled 1`] = `
Object {
  "expanded": Array [
    "panel2",
  ],
}
`;

exports[`Joke Reducer returns an initial state when passed undefined 1`] = `
Object {
  "expanded": Array [
    "panel1",
    "panel2",
  ],
  "favoriteJokes": Array [],
  "intervalActive": undefined,
  "jokes": Array [],
  "loading": true,
  "page": 0,
}
`;
